supervisor/homeassistant/websocket.py

```diff
@@ -22,61 +26,112 @@ class WSClient:
-        self.ha_version: AwesomeVersion = ha_version
-        self.client: aiohttp.ClientWebSocketResponse = client
-        self.message_id: int = 0
-        self._lock: asyncio.Lock = asyncio.Lock()
+        self.ha_version = ha_version
+        self._client = client
+        self._message_id: int = 0
+        self._loop = loop
+        self._futures: dict[int, asyncio.Future[dict]] = {}
+
+    @property
+    def connected(self) -> bool:
+        """Return if we're currently connected."""
+        return self._client is not None and not self._client.closed
+
+    async def close(self) -> None:
+        """Close down the client."""
+        if not self._client.closed:
+            await self._client.close()
+
+    async def async_send_command(self, message: dict[str, Any]) -> dict | None:
+        """Send a websocket message, and return the response."""
+        self._message_id += 1
+        message["id"] = self._message_id
+        self._futures[message["id"]] = self._loop.create_future()
+        _LOGGER.debug("Sending: %s", message)
+        try:
+            await self._client.send_json(message)
+        except ConnectionError as err:
+            raise HomeAssistantWSConnectionError(err) from err
 
-    async def async_send_command(self, message: dict[str, Any]):
-        """Send a websocket command."""
-        async with self._lock:
-            self.message_id += 1
-            message["id"] = self.message_id
+        try:
+            return await self._futures[message["id"]]
+        finally:
+            self._futures.pop(message["id"])
 
-            _LOGGER.debug("Sending: %s", message)
-            try:
-                await self.client.send_json(message)
-            except ConnectionError:
-                return
+    async def start_listener(self) -> None:
+        """Start listening to the websocket."""
+        if not self.connected:
+            raise HomeAssistantWSConnectionError("Not connected when start listening")
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
+
+        except HomeAssistantWSError as err:
+            _LOGGER.warning(err)
+
+        finally:
+            await self.close()
+
+    async def _receive_json(self) -> None:
+        """Receive json."""
+        msg = await self._client.receive()
+        _LOGGER.debug("Received: %s", msg)
 
-            _LOGGER.debug("Received: %s", response)
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
+
+        if msg.type != WSMsgType.TEXT:
+            raise HomeAssistantWSError(f"Received non-Text message: {msg.type}")
```

Comment:
This seems an error which which is related to the message itself, I think we should pass it to the caller via `future.set_exception(...)`.

Reply:
At this point we don't know if we have something waiting, and we can't know until we read the message data further down